crates/ty_python_semantic/src/module_name.rs

```diff
@@ -127,6 +127,60 @@ impl ModuleName {
         true
     }
 
+    /// Given a parent module name of this module name, return the relative
+    /// portion of this module name.
+    ///
+    /// For example, a parent module name of `importlib` with this module name
+    /// as `importlib.resources`, this returns `resources`.
+    ///
+    /// If `parent` isn't a parent name of this module name, then this returns
+    /// `None`.
+    ///
+    /// # Examples
+    ///
+    /// This example shows some cases where `parent` is an actual parent of the
+    /// module name:
+    ///
+    /// ```
+    /// use ty_python_semantic::ModuleName;
+    ///
+    /// let this = ModuleName::new_static("importlib.resources").unwrap();
+    /// let parent = ModuleName::new_static("importlib").unwrap();
+    /// assert_eq!(this.relative_to(&parent), ModuleName::new_static("resources"));
+    ///
+    /// let this = ModuleName::new_static("foo.bar.baz.quux").unwrap();
+    /// let parent = ModuleName::new_static("foo.bar").unwrap();
+    /// assert_eq!(this.relative_to(&parent), ModuleName::new_static("baz.quux"));
+    /// ```
+    ///
+    /// This shows some cases where it isn't a parent:
+    ///
+    /// ```
+    /// use ty_python_semantic::ModuleName;
+    ///
+    /// let this = ModuleName::new_static("importliblib.resources").unwrap();
+    /// let parent = ModuleName::new_static("importlib").unwrap();
+    /// assert_eq!(this.relative_to(&parent), None);
+    ///
+    /// let this = ModuleName::new_static("foo.bar.baz.quux").unwrap();
+    /// let parent = ModuleName::new_static("foo.barbaz").unwrap();
+    /// assert_eq!(this.relative_to(&parent), None);
+    ///
+    /// let this = ModuleName::new_static("importlibbbbb.resources").unwrap();
+    /// let parent = ModuleName::new_static("importlib").unwrap();
+    /// assert_eq!(this.relative_to(&parent), None);
+    /// ```
+    #[must_use]
+    pub fn relative_to(&self, parent: &ModuleName) -> Option<ModuleName> {
+        let relative_name = self.0.strip_prefix(&*parent.0)?.strip_prefix('.')?;
+        // At this point, `relative_name` *has* to be a
+        // proper suffix of `self`. Otherwise, one of the two
+        // `strip_prefix` calls above would return `None`.
+        // (Notably, a valid `ModuleName` cannot end with a `.`.)
+        assert!(!relative_name.is_empty());
```

Comment:
you could maybe also add a debug assertion encapsulating your explanation in https://github.com/astral-sh/ruff/pull/18898#discussion_r2162012481 of why it's correct to just rely on the `strip_prefix()` calls here:

```suggestion
        // At this point, `relative_name` *has* to be a
        // proper suffix of `self`. Otherwise, one of the two
        // `strip_prefix` calls above would return `None`.
        // (Notably, a valid `ModuleName` cannot end with a `.`.)
        assert!(!relative_name.is_empty());
        debug_assert!(self.starts_with(parent));
```

Reply:
Good point. Done. I thought this was a little paranoid, and I was inclined to resist since `strip_prefix` seems to make this vacuously true... but that's wrong. Because `starts_with` is smarter than a simple string prefix check. In theory, the implementations could diverge such that this assert does indeed trip.